chirho/counterfactual/handlers/explanation.py

```diff
@@ -0,0 +1,131 @@
+import contextlib
+from typing import Callable, Iterable, Mapping, Optional, ParamSpec, TypeVar
+
+import pyro
+import pyro.distributions
+import torch
+
+from chirho.counterfactual.handlers.counterfactual import BiasedPreemptions
+from chirho.counterfactual.handlers.selection import get_factual_indices
+from chirho.indexed.ops import IndexSet, cond, gather, indices_of, scatter
+from chirho.interventional.handlers import do
+from chirho.interventional.ops import Intervention
+from chirho.observational.handlers import condition
+from chirho.observational.handlers.condition import Factors
+from chirho.observational.ops import Observation
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def factual_preemption(
+    antecedents: Optional[Iterable[str]] = None, event_dim: int = 0
+) -> Callable[[T], T]:
+    def _preemption_with_factual(value: T) -> T:
+        if antecedents is None:
+            antecedents_ = list(indices_of(value, event_dim=event_dim).keys())
```

Comment:
Note the new default behavior of `factual_preemption` (previously `preempt_with_factual`) here. This is not exercised in the query code below or the pair of unit tests for `PartOfCause` initially included with this PR, and may be assuming too much about the free variables of `value`. I may remove or revert this before a version of `factual_preemption` lands in `master`.